File: src/game-bridge-client.js
import fs from 'fs/promises';
import path from 'path';

/**
 * Client for communicating with the Game Bridge HTTP API (running game)
 */
export class GameBridgeClient {
    constructor(host = '127.0.0.1', port = 8766) {
        this.baseUrl = `http://${host}:${port}`;
        this.timeout = 10000;
        this.editorClient = null;
    }

    /**
     * Set the editor client reference for debugger state checking
     * @param {EditorApiClient} client 
     */
    setEditorClient(client) {
        this.editorClient = client;
    }

    /**
     * Check if any debugger session is paused
     * @returns {Promise<{paused: boolean, sessions: Array}>}
     */
    async checkDebuggerPaused() {
        if (!this.editorClient) {
            return { paused: false, sessions: [] };
        }

        try {
            const sessions = await this.editorClient.getDebuggerSessions();
            const pausedSessions = Array.isArray(sessions)
                ? sessions.filter(s => s.paused)
                : [];
            return {
                paused: pausedSessions.length > 0,
                sessions: pausedSessions
            };
        } catch {
            // Editor might not be responding either
            return { paused: false, sessions: [] };
        }
    }

    async request(method, endpoint, data = null) {
        const options = {
            method,
            signal: AbortSignal.timeout(this.timeout)
        };

        if (data) {
            options.headers = { 'Content-Type': 'application/json' };
            options.body = JSON.stringify(data);
        }

        try {
            const response = await fetch(`${this.baseUrl}${endpoint}`, options);

            if (!response.ok) {
                throw new Error(`Game Bridge error: ${response.status} ${response.statusText}`);
            }

            return await response.json();
        } catch (error) {
            // Handle timeout - AbortError in older Node.js, TimeoutError in newer versions
            if (error.name === 'AbortError' || error.name === 'TimeoutError') {
                // Timeout occurred - check if game is paused in debugger
                const debuggerState = await this.checkDebuggerPaused();

                if (debuggerState.paused) {
                    const session = debuggerState.sessions[0]; // Primary session

                    let errorDetails = [];

                    // Session basic info
                    errorDetails.push(`Session ${session.id}: paused=${session.paused}, can_debug=${session.can_debug}`);

                    // Break reason and location
                    if (session.reason) {
                        errorDetails.push(`Reason: ${session.reason}`);
                    }
                    if (session.break_file) {
                        const location = session.break_line >= 0
                            ? `${session.break_file}:${session.break_line}`
                            : session.break_file;
                        errorDetails.push(`Location: ${location}`);
                    }

                    // Actual error message
                    if (session.error) {
                        errorDetails.push(`Error: ${session.error}`);
                    }

                    // Stack trace (if available)
                    let stackInfo = '';
                    if (session.stack && session.stack.length > 0) {
                        const stackLines = session.stack.map((frame, i) => {
                            const loc = frame.line >= 0 ? `${frame.file}:${frame.line}` : frame.file;
                            return `  ${i}: ${frame.function || '<anonymous>'} at ${loc}`;
                        }).join('\n');
                        stackInfo = `\n\nStack trace:\n${stackLines}`;
                    }

                    throw new Error(
                        `Game Bridge timeout - the game appears to be stopped in the debugger.\n\n` +
                        `${errorDetails.join('\n')}${stackInfo}\n\n` +
                        `To inspect or continue:\n` +
                        `1. Use godot_debugger_sessions to see full debugger state\n` +
                        `2. Use godot_debugger_resume to continue execution\n` +
                        `3. Or use godot_debugger_step_over to step through the code`
                    );
                }

                throw new Error(`Request timeout to Game Bridge at ${this.baseUrl}`);
            }
            throw error;
        }
    }

    // Screenshot from running game
    async captureScreenshot(options = {}) {
        const params = new URLSearchParams();
        if (options.maxWidth) params.set('max_width', options.maxWidth);
        if (options.maxHeight) params.set('max_height', options.maxHeight);
        // We force save_to_disk=false on the Godot side so we get the data back
        // and handle saving it on the Node.js side. (Godot would save to user://)

        const queryString = params.toString();
        const url = queryString ? `/screenshot?${queryString}` : '/screenshot';

        const result = await this.request('GET', url);

        // Check for 'data' (Game Bridge uses "data" key for base64) or 'base64' (legacy)
        const base64Data = result.data || result.base64;

        if (base64Data) {
            try {
                const buffer = Buffer.from(base64Data, 'base64');
                const filename = `screenshot_${Date.now()}.png`;
                const filepath = path.resolve(process.cwd(), filename);
                await fs.writeFile(filepath, buffer);

                // Return generic info + path, explicitly exclude base64
                return {
                    width: result.width,
                    height: result.height,
                    saved_to: filepath,
                    note: "Image saved to disk to avoid context overflow"
                };
            } catch (e) {
                console.error("Failed to save screenshot to disk:", e);
                // Fallback? Or just error.
                return { error: "Failed to save screenshot to disk", details: e.message };
            }
        }
        return result;
    }

    // Scene tree of running game
    async getSceneTree(maxDepth = 10) {
        return this.request('GET', `/scene_tree?max_depth=${maxDepth}`);
    }

    // Send single input event
    async sendInput(eventData) {
        return this.request('POST', '/input', eventData);
    }

    // Send input sequence
    async sendInputSequence(sequence, captureScreenshot = false) {
        return this.request('POST', '/input-sequence', {
            sequence,
            capture_screenshot: captureScreenshot
        });
    }

    // Convenience methods for common inputs
    async sendAction(actionName, pressed = true, strength = 1.0) {
        return this.sendInput({
            event_type: 'action',
            action_name: actionName,
            pressed,
            strength
        });
    }

    async sendKey(keycode, pressed = true, modifiers = {}) {
        return this.sendInput({
            event_type: 'key',
            keycode,
            pressed,
            ...modifiers
        });
    }

    async sendMouseButton(buttonIndex, x, y, pressed = true, doubleClick = false) {
        return this.sendInput({
            event_type: 'mouse_button',
            button_index: buttonIndex,
            position_x: x,
            position_y: y,
            pressed,
            double_click: doubleClick
        });
    }

    async sendMouseMotion(x, y, relativeX = 0, relativeY = 0) {
        return this.sendInput({
            event_type: 'mouse_motion',
            position_x: x,
            position_y: y,
            relative_x: relativeX,
            relative_y: relativeY
        });
    }

    // Health check
    async isGameRunning() {
        try {
            await fetch(`${this.baseUrl}/screenshot`, {
                method: 'HEAD',
                signal: AbortSignal.timeout(1000)
            });
            return true;
        } catch {
            return false;
        }
    }
}

File: src/godot_mcp/game_bridge_client.py
"""Client for communicating with the Game Bridge HTTP API (running game)."""

import base64
import sys
import time
from pathlib import Path

import httpx


def log(*args):
    print(*args, file=sys.stderr, flush=True)


class GameBridgeClient:
    def __init__(self, host: str = "127.0.0.1", port: int = 8766):
        self.base_url = f"http://{host}:{port}"
        self.timeout = 10.0
        self.editor_client = None

    def set_editor_client(self, client):
        """Editor client reference for debugger state checking on timeout."""
        self.editor_client = client

    async def check_debugger_paused(self) -> dict:
        if not self.editor_client:
            return {"paused": False, "sessions": []}
        try:
            sessions = await self.editor_client.get_debugger_sessions()
            paused = [s for s in sessions if s.get("paused")] if isinstance(sessions, list) else []
            return {"paused": bool(paused), "sessions": paused}
        except Exception:
            # Editor might not be responding either
            return {"paused": False, "sessions": []}

    async def request(self, method: str, endpoint: str, data: dict | None = None):
        try:
            async with httpx.AsyncClient(timeout=self.timeout) as client:
                response = await client.request(
                    method, f"{self.base_url}{endpoint}",
                    json=data if data is not None else None)
            if response.status_code >= 400:
                raise RuntimeError(
                    f"Game Bridge error: {response.status_code} {response.reason_phrase}")
            return response.json()
        except httpx.TimeoutException:
            # Timeout occurred - check if game is paused in debugger
            state = await self.check_debugger_paused()
            if state["paused"]:
                session = state["sessions"][0]  # Primary session
                details = [
                    f"Session {session.get('id')}: paused={session.get('paused')}, "
                    f"can_debug={session.get('can_debug')}"
                ]
                if session.get("reason"):
                    details.append(f"Reason: {session['reason']}")
                if session.get("break_file"):
                    line = session.get("break_line", -1)
                    location = (f"{session['break_file']}:{line}"
                                if line >= 0 else session["break_file"])
                    details.append(f"Location: {location}")
                if session.get("error"):
                    details.append(f"Error: {session['error']}")

                stack_info = ""
                if session.get("stack"):
                    frames = []
                    for i, frame in enumerate(session["stack"]):
                        loc = (f"{frame.get('file')}:{frame.get('line')}"
                               if frame.get("line", -1) >= 0 else frame.get("file"))
                        frames.append(f"  {i}: {frame.get('function') or '<anonymous>'} at {loc}")
                    stack_info = "\n\nStack trace:\n" + "\n".join(frames)

                raise TimeoutError(
                    "Game Bridge timeout - the game appears to be stopped in the debugger.\n\n"
                    + "\n".join(details) + stack_info + "\n\n"
                    "To inspect or continue:\n"
                    "1. godot_debugger_sessions shows the paused state and stack trace\n"
                    "2. godot_dap_stack_trace / godot_dap_scopes / godot_dap_variables "
                    "inspect frames and variables; godot_dap_evaluate runs expressions\n"
                    "3. godot_debugger_resume continues; godot_debugger_step_over / "
                    "godot_debugger_step_into step through the code"
                )
            raise TimeoutError(f"Request timeout to Game Bridge at {self.base_url}")

    # Screenshot from running game
    async def capture_screenshot(self, max_width: int | None = None,
                                 max_height: int | None = None,
                                 save_to_disk: bool = False):
        params = []
        if max_width:
            params.append(f"max_width={max_width}")
        if max_height:
            params.append(f"max_height={max_height}")
        # save_to_disk is handled on this side so the image lands in cwd
        # instead of Godot's user:// directory.
        url = "/screenshot" + (f"?{'&'.join(params)}" if params else "")

        result = await self.request("GET", url)

        # Game Bridge uses "data" key for base64; "base64" is legacy
        base64_data = result.get("data") or result.get("base64")
        if base64_data:
            try:
                filename = f"screenshot_{int(time.time() * 1000)}.png"
                filepath = Path.cwd() / filename
                filepath.write_bytes(base64.b64decode(base64_data))
                return {
                    "width": result.get("width"),
                    "height": result.get("height"),
                    "saved_to": str(filepath),
                    "note": "Image saved to disk to avoid context overflow",
                }
            except Exception as e:
                log(f"Failed to save screenshot to disk: {e}")
                return {"error": "Failed to save screenshot to disk", "details": str(e)}
        return result

    # Raw PNG bytes for MCP resource consumers (no disk round-trip)
    async def capture_screenshot_bytes(self, max_width: int = 1280,
                                       max_height: int = 720) -> bytes:
        result = await self.request(
            "GET", f"/screenshot?max_width={max_width}&max_height={max_height}")
        base64_data = result.get("data") or result.get("base64")
        if base64_data:
            return base64.b64decode(base64_data)
        # Oversized screenshots are written to disk by the game bridge
        file_path = result.get("file_path")
        if file_path:
            return Path(file_path).read_bytes()
        raise RuntimeError(f"Screenshot did not return image data: {result}")

    # Scene tree of running game
    async def get_scene_tree(self, max_depth: int = 10):
        return await self.request("GET", f"/scene_tree?max_depth={max_depth}")

    # Send single input event
    async def send_input(self, event_data: dict):
        return await self.request("POST", "/input", event_data)

    # Synthesize a full click (hover motion + press + release)
    async def click(self, node_path: str | None = None, x: float | None = None,
                    y: float | None = None, button_index: int = 1,
                    double_click: bool = False,
                    offset_x: float = 0.0, offset_y: float = 0.0):
        body: dict = {"button_index": button_index, "double_click": double_click}
        if node_path:
            body["node_path"] = node_path
            body["offset_x"] = offset_x
            body["offset_y"] = offset_y
        elif x is not None and y is not None:
            body["x"] = x
            body["y"] = y
        else:
            raise ValueError("Provide node_path or x/y coordinates")
        return await self.request("POST", "/click", body)

    # Send input sequence
    async def send_input_sequence(self, sequence: list, capture_screenshot: bool = False):
        return await self.request("POST", "/input-sequence", {
            "sequence": sequence,
            "capture_screenshot": capture_screenshot,
        })

    # Health check
    async def is_game_running(self) -> bool:
        try:
            async with httpx.AsyncClient(timeout=1.0) as client:
                await client.head(f"{self.base_url}/screenshot")
            return True
        except Exception:
            return False
